mods/features/preferredVideoQuality.js

```diff
@@ -0,0 +1,149 @@
+                if (obj && typeof obj.getAvailableQualityData === 'function') {
+                    acc.player = obj;
+                }
+                if (obj && obj.instance && typeof obj.instance.resolveCommand === 'function') {
+                    acc.app = obj.instance;
+                }
+                return acc;
+            }, { player: null, app: null });
+
+            if (player && app) {
+                let stringValue = null;
+                let appliedLabel = "Auto";
+
+                if (quality === 'auto') {
+                    stringValue = JSON.stringify({
+                        quality: "auto"
+                    });
+                }
+                else {
+                    const qualities = player.getAvailableQualityData();
+                    // console.log("[PreferredQuality] Available qualities:", qualities);
+
+                    if (qualities && qualities.length > 0) {
+                        const parseQuality = (q) => parseInt((q || '').toString().replace(/[^0-9]/g, ''), 10) || 0;
+                        const preferredNum = parseQuality(quality);
+
+                        const sortedQualities = qualities.map(q => ({
+                            original: q,
+                            val: parseQuality(q.qualityLabel)
+                        })).sort((a, b) => b.val - a.val);
+
+                        let target = sortedQualities.find(q => q.val <= preferredNum);
+
+                        if (!target) {
+                            target = sortedQualities[sortedQualities.length - 1];
+                        }
+
+                        if (target) {
+                            stringValue = JSON.stringify({
+                                quality: target.original.quality,
+                                formatId: target.original.formatId
+                            });
+                            appliedLabel = target.original.qualityLabel;
+                        }
+                    }
+                }
+
+                if (stringValue) {
+                    const command = {
+                        setClientSettingEndpoint: {
+                            settingDatas: [{
+                                clientSettingEnum: {
+                                    item: "PLAYBACK_QUALITY"
+                                },
+                                stringValue: stringValue
+                            }]
+                        }
+                    };
+                    app.resolveCommand(command);
+                    // console.log("[PreferredQuality] Applied quality:", appliedLabel);
+                }
+            }
+        } catch (e) {
+            console.warn('[PreferredQuality] Method failed:', e);
+        }
+    }
+}
+
+window.preferredVideoQualityHandler = null;
+
+window.addEventListener('hashchange', () => {
+    const newURL = new URL(location.hash.substring(1), location.href);
+    const videoID = newURL.search.replace('?v=', '').split('&')[0];
```

Comment:
From what I can see, this is entirely useless. If you need to check if the video is playing or not, listen to `onStateChange` event on the `.html5-video-player` element and do `videoPlayer.getPlayerStateObject().isPlaying`